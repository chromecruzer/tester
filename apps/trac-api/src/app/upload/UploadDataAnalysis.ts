import UploadStoreManagement from "../datastore/UploadStoreManagement";
import {UploadAnalysisA} from "./UploadAnalysisA";
import util from "util";

const dump = (obj, depth = null) => util.inspect(obj, {depth, colors: true});

export default class UploadDataAnalysis extends UploadAnalysisA {
  constructor(dataStore: UploadStoreManagement, schemaName, tableName) {
    super(dataStore, schemaName, tableName);

  }
  async comparisons(session) {
    const comparisons = await this.dataStore.retrieve(session);
    return comparisons ? {
      adds: comparisons.adds.length,
      exists: comparisons.exists.length,
    } : null;
  }
}
