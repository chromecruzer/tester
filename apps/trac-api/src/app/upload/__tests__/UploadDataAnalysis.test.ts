import UploadDataAnalysis from "../UploadDataAnalysis";

describe('UploadDataAnalysis', () => {
  let clientMock, analysis, originals, updated, modifies, dataStore;
  beforeEach(() => {
    clientMock = {
      query: jest.fn(),
    };
    dataStore = {
      retrieve: jest.fn()
    }
    clientMock.query.mockImplementation(() => Promise.resolve());
    analysis = new UploadDataAnalysis(dataStore, 'schema-name', 'table-name');
    originals = {
      same: {
        lot: 'lott1223'
      },
      modified: {
        lot: 'lott1225'
      },
      removed: {
        lot: 'lott1226'
      },
    };
    updated = [
      [
        'lot1223',
      ],
      [
        'lot1225',
      ],
      [
        'lot1227',
      ],
    ];
    modifies = [
      {
        new: {
          uuid: "modified",
          lot: "lot1225",
        },
        originalUuid: "modified",
      },
    ];
  });
  describe('#analytics', () => {
    it('should return the correct length calculations', () => {
      dataStore.retrieve.mockImplementation(() =>
        ({adds: updated, modifies, removedUuids: ['removed']}))
      return analysis.analytics().then(actual => {
        expect(actual).toMatchSnapshot();
      });
    });
  });
  describe('#details', () => {
    it('should return list of added, modified, and removed records', () => {
      dataStore.retrieve.mockImplementation(() =>
        ({adds: updated, modifies, removedUuids: ['removed']}));
      jest.spyOn(analysis, 'getOriginals').mockImplementation(() => [{uuid: 'removed'}]);
      jest.spyOn(analysis, 'getModifications').mockImplementation(() => modifies);
      return analysis.details(clientMock).then(actual => {
        expect(actual).toMatchSnapshot();
      });
    });
  });
  describe('#getOriginals', () => {
    it('should create correct sql to get records based on a list uuids', () => {
      clientMock.query.mockReturnValue(Promise.resolve({rows: []}));
      return analysis.getOriginals(['uuid01', 'uuid02', 'uuid03'], clientMock).then(() => {
        expect(clientMock.query.mock.calls[0][0]).toMatchSnapshot();
      });
    });
  });
  describe('#getModifications', () => {
    it('should fills in the modifies portion of comparisons correctly', () => {
      jest.spyOn(analysis, 'getOriginals').mockImplementation(() => {
        return Promise.resolve(originals);
      });
      return analysis.getModifications({modifies}, clientMock)
        .then(result => {
          expect(result).toMatchSnapshot();
        });
    });
  });
});
