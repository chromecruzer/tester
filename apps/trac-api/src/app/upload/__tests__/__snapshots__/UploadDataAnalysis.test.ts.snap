// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadDataAnalysis #analytics should return the correct length calculations 1`] = `
Object {
  "adds": 3,
  "modifies": 1,
  "removes": 1,
}
`;

exports[`UploadDataAnalysis #details should return list of added, modified, and removed records 1`] = `
Object {
  "adds": Array [
    Array [
      "lot1223",
    ],
    Array [
      "lot1225",
    ],
    Array [
      "lot1227",
    ],
  ],
  "modifies": Array [
    Object {
      "new": Object {
        "lot": "lot1225",
        "uuid": "modified",
      },
      "originalUuid": "modified",
    },
  ],
  "removes": Array [
    Object {
      "uuid": "removed",
    },
  ],
}
`;

exports[`UploadDataAnalysis #getModifications should fills in the modifies portion of comparisons correctly 1`] = `
Array [
  Object {
    "original": Object {
      "lot": "lott1225",
    },
    "update": Object {
      "lot": "lot1225",
      "uuid": "modified",
    },
  },
]
`;

exports[`UploadDataAnalysis #getOriginals should create correct sql to get records based on a list uuids 1`] = `"SELECT * FROM schema-name.table-name WHERE uuid IN ( 'uuid01', 'uuid02', 'uuid03');"`;
